source/common/tcp_proxy/upstream.h

```diff
@@ -68,16 +85,243 @@ class HttpConnPool : public GenericConnPool, public Http::ConnectionPool::Callba
                    Upstream::HostDescriptionConstSharedPtr host, StreamInfo::StreamInfo& info,
                    absl::optional<Http::Protocol>) override;
 
+  void onUpstreamHostSelected(Upstream::HostDescriptionConstSharedPtr);
+  void onHttpPoolReady(Upstream::HostDescriptionConstSharedPtr& host,
+                       Ssl::ConnectionInfoConstSharedPtr ssl_info);
+
+  struct NullHedgePolicy : public Router::HedgePolicy {
+    // Router::HedgePolicy
+    uint32_t initialRequests() const override { return 1; }
+    const envoy::type::v3::FractionalPercent& additionalRequestChance() const override {
+      return additional_request_chance_;
+    }
+    bool hedgeOnPerTryTimeout() const override { return false; }
+
+    const envoy::type::v3::FractionalPercent additional_request_chance_;
+  };
+
+  struct NullRateLimitPolicy : public Router::RateLimitPolicy {
+    // Router::RateLimitPolicy
+    const std::vector<std::reference_wrapper<const Router::RateLimitPolicyEntry>>&
+    getApplicableRateLimit(uint64_t) const override {
+      return rate_limit_policy_entry_;
+    }
+    bool empty() const override { return true; }
+
+    static const std::vector<std::reference_wrapper<const Router::RateLimitPolicyEntry>>
+        rate_limit_policy_entry_;
```

Comment:
What's the story with this static member? I'm confused by all of these Null structs and global variables. This seems  a bit scary to me.

Reply:
UpstreamRequest accepts `RouterFilterInterface`, so all these Null structs and global variables are to implement RouterFilterInterface which is passed to initialize Router::UpstreamRequest. Many of these will not be used in the tunneling flow. 
I looked at how async client is implemeting RouterFilterInterface, https://github.com/envoyproxy/envoy/blob/main/source/common/http/async_client_impl.h#L156 and following exactly same

Reply:
Instead of copying all of the NullFoo classes from async_client_impl.h, what would you think of moving those classes somewhere they can be shared?

Reply:
done.